File: backend/routing/mediaserver/get_library_sections.go
package routes_ms

import (
	"aura/config"
	"aura/logging"
	"aura/mediaserver"
	"aura/models"
	"aura/utils/httpx"
	"net/http"
)

type GetLibrarySections_Response struct {
	Sections []models.LibrarySection `json:"sections"`
}

// GetLibrarySections godoc
// @Summary      Get Library Sections
// @Description  Retrieve a list of library sections from the configured media server. This endpoint fetches the available library sections, including their ID, title, type, and path, allowing clients to display and interact with the media libraries configured on the media server.
// @Tags         MediaServer
// @Accept       json
// @Produce      json
// @Security 	 BearerAuth
// @Failure      401  {object}  httpx.UnauthorizedResponse "Unauthorized (only when Auth.Enabled=true)"
// @Success      200  {object}  httpx.JSONResponse{data=GetLibrarySections_Response}
// @Failure      500  {object}  httpx.JSONResponse "Internal Server Error"
// @Router       /api/mediaserver/libraries [get]
func GetLibrarySections(w http.ResponseWriter, r *http.Request) {
	ctx, ld := logging.CreateLoggingContext(r.Context(), r.URL.Path)
	logAction := ld.AddAction("Get Library Sections", logging.LevelInfo)
	ctx = logging.WithCurrentAction(ctx, logAction)
	var response GetLibrarySections_Response

	for _, libSection := range config.Current.MediaServer.Libraries {
		base := models.LibrarySectionBase{
			ID:    libSection.ID,
			Title: libSection.Title,
			Type:  libSection.Type,
			Path:  libSection.Path,
		}
		librarySection := models.LibrarySection{
			LibrarySectionBase: base,
		}

		found, Err := mediaserver.GetLibrarySectionDetails(ctx, &librarySection)
		if Err.Message != "" || !found {
			continue
		}
		response.Sections = append(response.Sections, librarySection)
	}

	if len(response.Sections) == 0 {
		logAction.SetError("No library sections found", "No valid library sections could be found on the configured Media Server", nil)
		httpx.SendResponse(w, ld, response)
		return
	}
	httpx.SendResponse(w, ld, response)
}

File: frontend/src/services/mediaserver/get-library-sections.ts
import apiClient from "@/services/api-client";
import { ReturnErrorMessage } from "@/services/api-error-return";

import { log } from "@/lib/logger";

import type { APIResponse } from "@/types/api/api-response";
import type { LibrarySection } from "@/types/media-and-posters/media-item-and-library";

export interface GetLibrarySections_Response {
  sections: LibrarySection[];
}

export const GetLibrarySections = async (): Promise<APIResponse<GetLibrarySections_Response>> => {
  log("INFO", "API - Media Server", "Fetch Library Sections", "Fetching all library sections");
  try {
    const response = await apiClient.get<APIResponse<GetLibrarySections_Response>>(`/mediaserver/libraries/`);
    if (response.data.status === "error") {
      throw new Error(response.data.error?.message || "Unknown error fetching all library sections");
    } else {
      log("INFO", "API - Media Server", "Fetch Library Sections", `Fetched all library sections successfully`);
    }
    return response.data;
  } catch (error) {
    log(
      "ERROR",
      "API - Media Server",
      "Fetch Library Sections",
      `Failed to fetch all library sections: ${error instanceof Error ? error.message : "Unknown error"}`,
      error
    );
    return ReturnErrorMessage<GetLibrarySections_Response>(error);
  }
};
